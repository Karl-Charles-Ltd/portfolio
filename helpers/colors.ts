/**
 * Get the hex color codes that are used throughout the project.
 * These colors and names should match with
 * assets/scss/settings/_settings.colors.scss.
 *
 * @param {String} colorName - The color name from the theme
 * @returns {String} - The hex code for the matching color
 */
export function colorFromTheme(colorName) {
  const colors = {
    primary: '#370b3c',
    secondary: '#6b2f93',
    tertiary: '#D0438F',
    quaternary: '#f40ee3',
    quinary: '#09ECFD',
    'quinary-light': '#1A5C7E',
    'quinary-dark': '#010226',
    black: '#000000',
    background: '#202227',
    charcoal: '#4D4D4D',
    'grey-one': '#949494',
    'grey-two': '#CDCDCD',
    'light-grey': '#E2E2E2',
    'light-grey-one': '#E2E2E2',
    'light-grey-two': '#F9F9F9',
    white: '#FFFFFF',
  };

  return colors[colorName];
}

export const themeMap = {
  primary: {
    background: 'primary',
    foreground: 'secondary',
  },
  'primary-dark': {
    background: 'primary-dark',
    foreground: 'secondary',
  },
  secondary: {
    background: 'secondary',
    foreground: 'white',
    accent: 'tertiary',
  },
  'secondary-dark': {
    background: 'secondary-dark',
    foreground: 'white',
  },
  'secondary-light': {
    background: 'secondary-light',
    foreground: 'white',
  },
  tertiary: {
    background: 'tertiary',
    foreground: 'secondary',
  },
  quaternary: {
    background: 'quaternary',
    foreground: 'secondary',
  },
  quinary: {
    background: 'quinary',
    foreground: 'white',
  },
  'quinary-darker': {
    background: 'quinary-darker',
    foreground: 'white',
  },
  white: {
    background: 'white',
    foreground: 'secondary',
  },
  charcoal: {
    background: 'charcoal',
    foreground: 'white',
  },
  'light-grey-one': {
    background: 'light-grey-one',
    foreground: 'secondary',
  },
};
