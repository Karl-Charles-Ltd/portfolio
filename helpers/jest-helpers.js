import { shallowMount, createLocalVue } from '@vue/test-utils';
import { createRenderer } from 'vue-server-renderer';
import VueRouter from 'vue-router';

export function setup() {
  const renderer = createRenderer();
  const localVue = createLocalVue();
  const router = new VueRouter();

  // setupPlugins(useVuex, localVue);

  return {
    renderer,
    router,
    localVue,
    shallowMount,
  };
}

export function componentFactory(component, options) {
  return shallowMount(component, options);
}

export function createSnapshot(component, options) {
  const wrapper = componentFactory(component, options);
  expect(wrapper.exists()).toBeTruthy();
  expect(wrapper.html()).toMatchSnapshot();
}
