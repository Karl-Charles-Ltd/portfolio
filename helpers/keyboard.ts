export function addKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.addEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}

export function removeKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.removeEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}
