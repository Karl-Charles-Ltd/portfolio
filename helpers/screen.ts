import * as screenBreakpoints from '~/constants/screen-breakpoints';

export const breakpoints: string[] = [
  screenBreakpoints.MOBILE,
  screenBreakpoints.TABLET,
  screenBreakpoints.SMALL_DESK,
  screenBreakpoints.LARGE_DESK,
  screenBreakpoints.HUGE_DESK,
];

interface Screen {
  [key: string]: string | null;
}

export default function mq({ from = null, until = null }: Screen, currentScreenSize: string): boolean {
  const currentBreakpointIndex = breakpoints.indexOf(currentScreenSize);

  if (from !== '' && from !== null) {
    const fromBreakpointIndex = breakpoints.indexOf(from || '');
    if (fromBreakpointIndex > currentBreakpointIndex) {
      return false;
    }
  }

  if (until !== '' && until !== null) {
    const untilBreakpointIndex = breakpoints.indexOf(until || '');
    if (untilBreakpointIndex <= currentBreakpointIndex) {
      return false;
    }
  }

  return true;
}
