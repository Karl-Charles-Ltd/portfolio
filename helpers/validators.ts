import { themeMap } from './colors';

export function validateTheme(theme: string, extraColors: string[] = []): boolean {
  const baseThemes: string[] = Object.keys(themeMap);
  const availableThemes: string[] = baseThemes.concat(extraColors);

  return availableThemes.includes(theme);
}
