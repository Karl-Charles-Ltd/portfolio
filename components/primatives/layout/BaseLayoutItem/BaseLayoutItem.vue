<template>
  <div :class="layoutItemClasses" @mouseover="$emit('mouseover')" @mouseleave="$emit('mouseleave')">
    <slot />
  </div>
</template>

<script>
import { breakpoints as availableBreakpoints } from '~/helpers/screen';
/**
 * Base Layout Item
 *
 * The Base Layout Item must only ever be used as a direct descendent of a BaseLayout.
 *
 * By default, it spans 100% width of the layout it's placed in.
 * You can alter this defauly behaviour by passing a string to the r prop, stating a ratio of column / max columns.
 *
 * <BaseLayout>
 *   <BaseLayoutItem r="1/2">   <!-- This will span one half of the width of the layout -->
 *     Content
 *   </BaseLayoutItem>
 * </BaseLayout>
 *
 * You can also change this on a per breakpoint basis, by using the breakpoint name as a prop:
 *
 * <BaseLayout>
 *   <!-- This will span one half by default, one third on tablet up, and one quarter on desktopSmall and above -->
 *   <BaseLayoutItem r="1/2" tablet="1/3" desktopSmall="1/4">
 *     Content
 *   </BaseLayoutItem>
 * </BaseLayout>
 */

const breakpointProps = availableBreakpoints.reduce(
  (props, breakpoint) => ({
    ...props,
    [breakpoint]: { type: String, default: null },
  }),
  {}
);

export default {
  name: 'BaseLayoutItem',
  props: {
    r: { type: String, default: '1/1' },
    ...breakpointProps,
  },
  computed: {
    layoutItemClasses() {
      return {
        'v-layout__item': true,
        [`u-${this.r}`]: this.r,
        ...availableBreakpoints.reduce(
          (breakpoints, breakpoint) => ({
            ...breakpoints,
            [`u-${this[breakpoint]}@${breakpoint}`]: this[breakpoint],
          }),
          {}
        ),
      };
    },
  },
};
</script>
