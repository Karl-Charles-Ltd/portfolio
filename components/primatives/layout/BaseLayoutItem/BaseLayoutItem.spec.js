import BaseLayoutItem from './BaseLayoutItem.vue';
import { setup, componentFactory, createSnapshot } from '~/helpers/jest-helpers';

const { localVue } = setup();

const shallowMountComponent = (options) => componentFactory(BaseLayoutItem, { localVue, ...options });

describe('BaseLayoutItem', () => {
  test('mounts and renders properly', () => {
    createSnapshot(BaseLayoutItem);
  });

  it(`has the width utility class if a ratio (r) prop is set`, () => {
    const ratio = '1/2';
    const wrapper = shallowMountComponent({ propsData: { r: ratio } });
    expect(wrapper.classes().includes(`u-${ratio}`)).toBeTruthy();
  });

  it(`has a default 100% width utility class if the ratio (r) prop is not set`, () => {
    const wrapper = shallowMountComponent();
    expect(wrapper.classes().includes('u-1/1')).toBeTruthy();
  });
});
