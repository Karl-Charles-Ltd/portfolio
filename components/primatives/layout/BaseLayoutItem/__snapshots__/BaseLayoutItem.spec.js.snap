// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseLayoutItem mounts and renders properly 1`] = `"<div class=\\"v-layout__item u-1/1\\"></div>"`;
